File: python/batch.py
import aerospike
from aerospike_helpers import expressions as exp
from aerospike_helpers.operations import expression_operations, map_operations, operations
from aerospike_helpers.batch import records as br

config = {
    "hosts": [("localhost", 3100), ("localhost", 3101), ("localhost", 3102)],
    "use_services_alternate": True
}
client = aerospike.client(config).connect()

# An example that will always return true
expr = exp.GT(2, 1).compile()

# Create a new batch policy
batch_policy = {'expressions': expr}

# Create the batch write policy
batch_write_policy = {'expressions': expr}

# Exists - Creates an array of ten keys and checks for their existance in the database

keys = []
for i in range(4995, 5006):
    batch_key = ('test', 'a_set', i)
    keys.append(batch_key)

# Check if record exists
exists = client.exists_many(keys)

# Access the records
for record in exists:
    (key, meta) = record
    if meta == None:
        print('Key: ', key[2], ' does not exist\\n')

# Create batch of keys
keys = []
for i in range(1,11):
    batch_key = ('test', 'ufodata', i)
    keys.append(batch_key)

# Read each whole record 
records = client.get_many(keys);

# Or specifiy bins 
# records = client.select_many(keys, ("report", "location"));

# Access the records 
for record in records:
    (key, meta, bins) = record
    # Do something
    print('Record: ', bins, '\\n')

client.close()

File: java/my-app/src/main/java/com/mycompany/app/Batch.java
package com.mycompany.app;

import com.aerospike.client.AerospikeClient;
import com.aerospike.client.BatchDelete;
import com.aerospike.client.BatchRead;
import com.aerospike.client.BatchRecord;
import com.aerospike.client.BatchResults;
import com.aerospike.client.BatchWrite;
import com.aerospike.client.Bin;
import com.aerospike.client.Key;
import com.aerospike.client.Operation;
import com.aerospike.client.Record;
import com.aerospike.client.Value;
import com.aerospike.client.cdt.MapOperation;
import com.aerospike.client.cdt.MapPolicy;
import com.aerospike.client.cdt.MapReturnType;
import com.aerospike.client.exp.Exp;
import com.aerospike.client.exp.ExpOperation;
import com.aerospike.client.exp.ExpReadFlags;
import com.aerospike.client.exp.ExpWriteFlags;
import com.aerospike.client.exp.Expression;
import com.aerospike.client.exp.ListExp;
import com.aerospike.client.exp.MapExp;
import com.aerospike.client.policy.BatchPolicy;
import com.aerospike.client.policy.BatchWritePolicy;
import com.aerospike.client.policy.ClientPolicy;

import java.util.Arrays;
import java.util.List;

public class Batch {
    public static void main(String[] args) {
        
        ClientPolicy clientPolicy = new ClientPolicy();
        clientPolicy.useServicesAlternate = true;
        AerospikeClient client = new AerospikeClient(clientPolicy, "localhost", 3100);
        
        // Batch Policy
        BatchPolicy batchPolicy = new BatchPolicy();
        batchPolicy.filterExp = Exp.build(
            // Always returns true, 2 > 1
            Exp.gt(Exp.val(2), Exp.val(1))
        );

        // Batch Write Policy
        BatchWritePolicy batchWritePolicy = new BatchWritePolicy();
        batchWritePolicy.filterExp = Exp.build(
            Exp.gt(Exp.val(2), Exp.val(1))
        );

        // Batch write on 10 keys
        Key[] write_keys = new Key[10];
        for (int i = 0; i < write_keys.length; i++) {
            write_keys[i] = new Key("test", "test", (i + 1));
        }

        // Define Operation Expressions
        Expression exp = Exp.build(
            MapExp.put(MapPolicy.Default, Exp.val("recent"), 
                Exp.and(
                    Exp.gt(Exp.intBin("occurred"), Exp.val(20211231)),
                    Exp.binExists("posted")
                ),
                Exp.mapBin("report")
            )
        );

        // Execute the write operation and return the report bin
        BatchResults batchResult = client.operate(batchPolicy, batchWritePolicy, write_keys,
            ExpOperation.write("report", exp, ExpWriteFlags.DEFAULT),
            Operation.get("report")
        );

        // Access the records 
        for (BatchRecord batchRecord : batchResult.records){
            Record record = batchRecord.record;
            if(record != null){
                // Do something
                System.out.format("Record: %s\\n", record.bins); 
            }
        }

        // Batch exists on 10 keys
        Key[] keys = new Key[10];
        for (int i = 0; i < 10; i++) {
            keys[i] = new Key("test", "test", i + 1);
        }
        boolean[] exists = client.exists(batchPolicy, keys);

        for (int i = 0; i < exists.length; i++) {
            System.out.format("Key: %s does not exist\n", keys[i].userKey);
        }

        // Batch reads, returning whole record or specified report/location bins
        Key[] read_keys = new Key[10];
        for (int i = 0; i < read_keys.length; i++) {
            keys[i] = new Key("test", "test", (i + 1));
        }
        Record[] read_records = client.get(batchPolicy, read_keys);
        // Record[] read_records = client.get(batchPolicy, read_keys, "report", "location");
        for (Record record : read_records) {
            if(record != null) {
                System.out.format("Record: %s\n", record.bins);
            }
        }
        client.close();
    }
}
